### PR TITLE
Add option to delete beatmaps

### LINKED ISSUE #3459: Make it possible to resurrect and delete graveyarded beatmaps
old:
chrome_2018-07-05_21-00-05

current:
chrome_2018-07-05_21-00-50
I think you could add two buttons next to the download button, only the creator should be able to see them.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/60527965-make-it-possible-to-resurrect-and-delete-graveyarded-beatmaps?utm_campaign=plugin&utm_content=tracker%2F21853994&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F21853994&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by notbakaneko**
closes #3459 

Also add deleting beatmapsets in general

---

### CODE REVIEW

File: app/Libraries/OsuAuthorize.php [hunk 1]
```diff
@@ -327,6 +327,17 @@ public function checkBeatmapDiscussionPostStore($user, $post)
         return 'ok';
     }
 
+    public function checkBeatmapsetDelete($user, $beatmapset)
+    {
+        $this->ensureLoggedIn($user);
+
+        if (!($user->isGMT() || $user->isQAT() || $user->getKey() === $beatmapset->user_id)) {
```
[the remark this answers is not in the document]
Comment (cl8n) [reply app/Libraries/OsuAuthorize.php]: scratch that, needs to be checked in `BeatmapsetDelete` so nobody can bypass

File: app/Libraries/OsuAuthorize.php [hunk 2]
```diff
@@ -327,6 +327,23 @@ public function checkBeatmapDiscussionPostStore($user, $post)
...
 
+    public function checkBeatmapsetDelete($user, $beatmapset)
+    {
+        $this->ensureLoggedIn($user);
+
+        if ($beatmapset->isGraveyard() && $user->getKey() === $beatmapset->user_id) {
+            return 'ok';
+        }
+
+        // same approved state as beatmaps that can be loved.
```
Comment (nanaya) [inline app/Libraries/OsuAuthorize.php]: not sure I get what it means.

File: app/Libraries/OsuAuthorize.php [hunk 3]
```diff
@@ -327,6 +327,23 @@ public function checkBeatmapDiscussionPostStore($user, $post)
...
+        // TODO: method should be renamed.
+        if (!$beatmapset->isScoreable() && ($user->isGMT() || $user->isQAT())) {
+            return 'ok';
+        }
+
+        return 'unauthorized';
```
Comment (nanaya) [inline app/Libraries/OsuAuthorize.php]: already the default if null

File: app/Models/Beatmapset.php [hunk 4]
```diff
@@ -952,6 +958,25 @@ public function afterCommit()
...
+            $this->beatmaps()->delete();
+
+            return parent::delete();
+        });
+    }
+
+    public function restore()
+    {
+        return $this->getConnection()->transaction(function () {
+            $this->beatmaps()->restore();
```
Comment (nanaya) [inline app/Models/Beatmapset.php]: looks like it should be `allBeatmaps`

File: app/Models/Beatmapset.php [hunk 5]
```diff
@@ -952,6 +958,25 @@ public function afterCommit()
...
+
+            return parent::restore();
```
Comment (notbakaneko) [inline app/Models/Beatmapset.php]: This actually does something terrible 🥇

File: app/Jobs/BeatmapsetDelete.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            );
+
+            $post = $this->beatmapset->getPost();
+            if ($post !== null) {
+                $post->post_text = 'This map has been deleted on the request of its creator. It is no longer available.';
+                $post->skipBeatmapPostRestrictions()->saveOrExplode();
+            }
+
+            if ($this->beatmapset->delete()) {
+                Storage::disk(config('osu.beatmapset.storage'))->delete($this->beatmapset->getKey());
```
[the remark this answers is not in the document]
Comment (nanaya) [reply app/Jobs/BeatmapsetDelete.php]: it's part of the model though. There's even filesize column, and storage of the file depends on the model's id.

Alternatively there's callback thingy which can be used.

[the remark this answers is not in the document]
Comment (nanaya) [reply app/Jobs/BeatmapsetDelete.php]: I'm thinking of it more of a state (visibility) change which side effect includes removal of the file.

[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Jobs/BeatmapsetDelete.php]: I don't think a visibility change should result in removing the file.

**Pull request merged**